src/analysis/monotone-analyzer-impl.h

```diff
@@ -0,0 +1,126 @@
+#ifndef wasm_analysis_monotone_analyzer_impl_h
+#define wasm_analysis_monotone_analyzer_impl_h
+
+#include <iostream>
+#include <unordered_map>
+
+#include "monotone-analyzer.h"
+
+namespace wasm::analysis {
+template<size_t N>
+BlockState<N>::BlockState(const BasicBlock* underlyingBlock)
+  : index(underlyingBlock->getIndex()), cfgBlock(underlyingBlock),
+    states(underlyingBlock->size() + 1, BitsetPowersetLattice<N>::getBottom()),
```

Comment:
I would expect that we would only store the single entry state for each block, rather than a separate state for each instruction in the block. After the analysis completes, the follow-up code can collect a separate state for each instruction if it needs to, but that shouldn't generally be necessary. For example, consider a live variable analysis, where storing intermediate states for instructions that don't involve locals should never be useful.

Reply:
I implemented this, but now that we don't store intermediate results, is there any situation where this now ignores the process of applying the LUB between two intermediate states? For example, suppose we have two intermediate states, A, and B. We have just finished calculating A and will now set B = LUB(A, B). If B is a superset of A, then LUB(A, B) = B, but this wouldn't happen if we don't store intermediate states, because we just set B to A.

Reply:
In the iterative worklist algorithm, the join doesn't take the LUB of the current state with the inputs, so my original questions was incorrect. 

Each join only takes the LUB of all states it depends upon. We can take the LUB between a basic block's input state and some input multiple times to update it efficiently (only changed values are used to update the LUB) because the transfer function isn't actually applied on the input state (it is only applied to get the next state). 

This means we don't have to worry about the transfer function being applied on some expression A affecting JOIN(A) and trying to maintain that state.